include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -78,6 +80,148 @@ __find_start_point(const _Rng1& __rng1, const _Rng2& __rng2, const _Index __i_el
     }
 }
 
+//Searching for an intersection of a merge matrix (n1, n2) diagonal with the Merge Path to define sub-ranges
+//to serial merge. For example, a merge matrix for [0,1,1,2,3] and [0,0,2,3] is shown below:
+//     0   1  1  2   3
+//    ------------------
+//   |--->
+// 0 | 0 | 1  1  1   1
+//   |   |
+// 0 | 0 | 1  1  1   1
+//   |   ---------->
+// 2 | 0   0  0  0 | 1
+//   |             ---->
+// 3 | 0   0  0  0   0 |
+template <typename _Rng1, typename _Rng2, typename _Index, typename _Compare>
+_split_point_t<_Index>
+__find_start_point_in(const _Rng1& __rng1, const _Index __rng1_from, _Index __rng1_to, const _Rng2& __rng2,
+                      const _Index __rng2_from, _Index __rng2_to, const _Index __i_elem, _Compare __comp)
+{
+    assert(__rng1_from <= __rng1_to);
+    assert(__rng2_from <= __rng2_to);
+
+    assert(__rng1_to > 0 || __rng2_to > 0);
+
+    if constexpr (!std::is_pointer_v<_Rng1>)
+        assert(__rng1_to <= __rng1.size());
+    if constexpr (!std::is_pointer_v<_Rng2>)
+        assert(__rng2_to <= __rng2.size());
+
```

Comment:
The lines 105-108 look suspicious by several reasons:
1) it is a mix compile time and runtime conditions.
2) Rng1 (Rng2) is not supposed to be a pointer type. If we want to check that we can add a relevant `static_assert` here.
3) If we want to have some bound checks for rng1 and rng2, I would suggest writing that in a couple of asserts above (lines 100-102)  like that:
```
assert(0 <= __rng1_from && __rng1_from < __rng1_to && __rng1_to < __rng1.size());
assert(0 <= __rng2_from && __rng2_from < __rng2_to && __rng2_to < __rng2.size());
```

Reply:
But in the test `sort.pass` we has the case when we pass into this function pointers as `__rng1` and `__rng2`.
So I propose to rewrite:
```C++
    if constexpr (!std::is_pointer_v<_Rng1> && !std::is_pointer_v<_Rng2>)
    {
        assert(0 <= __rng1_from && __rng1_from < __rng1_to && __rng1_to < __rng1.size());
        assert(0 <= __rng2_from && __rng2_from < __rng2_to && __rng2_to < __rng2.size());
    }
    else
    {
        assert(0 <= __rng1_from && __rng1_from < __rng1_to);
        assert(0 <= __rng2_from && __rng2_from < __rng2_to);
    }
```

Reply:
Well. So, the function can accept some type `_Rng1` which is an "accessible" or, other words the type must have subscription operator `operator[]`, nothing more.
Basically,  the function code doesn't need `size()` "method". And I think we should not demand it for the sake of "assert".

Reply:
For debug purposes it's very useful, but I am ready to remove `.size()` calls. Will remove?

Reply:
If we have full control over all the types of ranges that can be passed as inputs to the code, and so we can ensure that the `size` method is always available, then it would be OK to use it for assertions/debug.

If we do not have such level of control, and ranges can be of user-specified types, then we clearly cannot use this method.

Reply:
These checks has been removed at all.